src/hotspot/share/opto/vectorIntrinsics.cpp

```diff
@@ -698,9 +698,6 @@ bool LibraryCallKit::inline_vector_mask_operation() {
   ciKlass* mbox_klass = mask_klass->const_oop()->as_instance()->java_lang_Class_klass();
   const TypeInstPtr* mask_box_type = TypeInstPtr::make_exact(TypePtr::NotNull, mbox_klass);
   Node* mask_vec = unbox_vector(mask, mask_box_type, elem_bt, num_elem, true);
-  if (mask_vec->bottom_type()->isa_vectmask() == NULL) {
-    mask_vec = gvn().transform(VectorStoreMaskNode::make(gvn(), mask_vec, elem_bt, num_elem));
-  }
```

Comment:
There may have some regressions on AArch64 after refactoring those reduction operations base on this PR. 

With the `VectorStoreMaskNode`, the load-store pair could be optimized. After removing that, `VectorLoadMaskNode` needs extra extension instructions for various type. E.g.

```
After:
        ldr     h16, [x10, #16]  // LoadVector
        uxtl    v16.8h, v16.8b
        uxtl    v16.4s, v16.4h
        uxtl    v16.2d, v16.2s
        neg     v16.2d, v16.2d  // VectorLoadMask
        neg     v17.2d, v16.2d
        addv    b17, v17.16b
        umov    w11, v17.b[0]   // TrueCount
        sxtw    x0, w11         // StoreVector

Before:

        ldr     h16, [x10, #16] // LoadVector
        addv    b17, v16.8b
        umov    w11, v17.b[0]   // TrueCount
        sxtw    x0, w11         // StoreVector
```

No sliver bullet on this problem. Either you can match `VectorMaskTrueCount (VectorLoadMask src)` on AArch64 with this mid-end change in your patch, or just let the mid-end stay the same and match `VectorMaskTrueCount (VectorStoreMask src)` on x86 to pursue the better performance. The later would be better for its less effort.

Reply:
Thank you very much for your investigation, I have reverted the changes in the mid-end and pushed the change to the back-end as your suggestion. Another approach may be to elide the `VectorLoadMaskNode` attached to a mask reduction node but I think it should be another PR later.